src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php

```diff
@@ -95,22 +132,66 @@ public function __construct(
    *   Max depth to search in.
-   * @param \Drupal\Core\Cache\RefinableCacheableDependencyInterface $metadata
-   *   Refinable cacheability metadata.
+   * @param string|null $language
+   *   Optional. Language to be respected for retrieved entities.
+   * @param bool $access
+   *   Whether check for access or not. Default is true.
+   * @param \Drupal\Core\Session\AccountInterface|null $accessUser
+   *   User entity to check access for. Default is null.
+   * @param string $accessOperation
+   *   Operation to check access for. Default is view.
+   * @param \Drupal\graphql\GraphQL\Execution\FieldContext $context
+   *   The caching context related to the current field.
    *
-   * @return \object[]
-   *   A list of stdClass terms in the given vocabulary.
+   * @return \GraphQL\Deferred|null
+   *   A promise that will return entities or NULL if there aren't any.
    */
-  public function resolve(string $vid, int $parent = 0, ?int $max_depth = NULL, RefinableCacheableDependencyInterface $metadata): array {
+  public function resolve(string $vid, ?int $parent, ?int $max_depth, ?string $language, ?bool $access, ?AccountInterface $accessUser, ?string $accessOperation, FieldContext $context): array {
     if (!isset($max_depth)) {
       $max_depth = self::MAX_DEPTH;
     }
-    // @todo This should use a buffer system similar to other entities are using.
+
     $terms = $this->entityTypeManager
       ->getStorage('taxonomy_term')
       ->loadTree($vid, $parent, $max_depth);
 
-    return $terms;
+    $term_ids = array_column($terms, 'tid');
+    $resolver = $this->entityBuffer->add('taxonomy_term', $term_ids);
+
+    return new Deferred(function () use ($language, $resolver, $context, $access, $accessUser, $accessOperation) {
+      /** @var \Drupal\Core\Entity\EntityInterface[] $entities */
+      if (!$entities = $resolver()) {
+        // If there is no entity with this id, add the list cache tags so that
+        // the cache entry is purged whenever a new entity of this type is
+        // saved.
+        $type = $this->entityTypeManager->getDefinition('taxonomy_term');
+        /** @var \Drupal\Core\Entity\EntityTypeInterface $type */
+        $tags = $type->getListCacheTags();
+        $context->addCacheTags($tags);
+        return [];
+      }
+
+      foreach ($entities as $id => $entity) {
+        $context->addCacheableDependency($entities[$id]);
+
+        if (isset($language) && $language !== $entities[$id]->language()->getId() && $entities[$id] instanceof TranslatableInterface) {
+          $entities[$id] = $entities[$id]->getTranslation($language);
+          $entities[$id]->addCacheContexts(["static:language:{$language}"]);
+        }
+
+        if ($access) {
+          /* @var $accessResult \Drupal\Core\Access\AccessResultInterface */
+          $accessResult = $entity->access($accessOperation, $accessUser, TRUE);
+          $context->addCacheableDependency($accessResult);
+          if ($accessResult->isForbidden()) {
```

Comment:
Has to be ->isAllowed() as @klausi just found in the other issue. This needs to b e fixed everywhere in our current code base. Maybe you can open a PR that just replaces ALL occurrences of `isForbidden` with `isAllowed` (and negating the condition) for the entire code base?

Reply:
Adjusted. Should be ready for review